src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,338 @@
+package ui.components.pickers;
+
+import backend.resource.TurboLabel;
+import util.Utility;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to represent the state of the label picker. In additional to representing a state,
+ * it contains the logic which handles toggling/highlighting of labels and simplifies the retrieval of
+ * the resulting status for UI to display.
+ */
+public class LabelPickerState {
+    Set<String> initialLabels;
+    List<String> addedLabels;
+    List<String> removedLabels;
+    List<String> matchedLabels;
+    OptionalInt currentSuggestionIndex;
+
+    /*
+     * Constructors
+     */
+
+    public LabelPickerState(Set<String> initialLabels) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = new ArrayList<>();
+        this.removedLabels = new ArrayList<>();
+        this.matchedLabels = new ArrayList<>();
+        this.currentSuggestionIndex = OptionalInt.empty();
+    }
+
+    private LabelPickerState(Set<String> initialLabels,
+                             List<String> addedLabels,
+                             List<String> removedLabels,
+                             List<String> matchedLabels,
+                             OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    private LabelPickerState(Set<String> initialLabels,
+                             List<String> addedLabels,
+                             List<String> removedLabels,
+                             List<String> matchedLabels) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.currentSuggestionIndex = OptionalInt.empty();
+    }
+
+    /**
+     * Determines resulting state based on given initial state and user input
+     * @param initialState
+     * @param repoLabels
+     * @param userInput 
+     * @return new state that corresponds with the user input
+     */
+    public static LabelPickerState determineState(LabelPickerState initialState, Set<String> repoLabels,
+                                                  String userInput) {
+        LabelPickerState state = initialState;
+        List<String> confirmedKeywords = getConfirmedKeywords(userInput);
+        for (String confirmedKeyword : confirmedKeywords) {
+            state = state.toggleLabel(repoLabels, confirmedKeyword);
+        }
+
+        Optional<String> keywordInProgess = getKeywordInProgress(userInput);
+        if (keywordInProgess.isPresent()) {
+            state = state.updateMatchedLabels(repoLabels, keywordInProgess.get());
+        }
+
+        return state;
```

Comment:
I feel that this should be a mutator method rather than a static method. The code could look like this:
``` java
LabelPickerState state = new LabelPickerState(repoLabels, issueLabels);
state.update(userInput);
```
And the `update` method can be like this:
```
split user input at spaces (without removing spaces)
for each word,
  if word ends in space, update state for confirmed word
  else update state for word-in-progress
```